javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/ReferenceCollector.java

```diff
@@ -201,6 +203,73 @@ private void collectHelperClasses(
     return references;
   }
 
+  public void prune() {
+    Set<Reference> helperClassesWithLibrarySuperType = getHelperClassesWithLibrarySuperType();
+
+    Set<Reference> needToKeepFieldsAndMethods = new HashSet<>();
+    for (Reference reference : helperClassesWithLibrarySuperType) {
+      addSuperClasses(reference.getClassName(), needToKeepFieldsAndMethods);
+    }
+
+    for (Iterator<Map.Entry<String, Reference>> i = references.entrySet().iterator();
+        i.hasNext(); ) {
+      Reference reference = i.next().getValue();
+      if (instrumentationClassPredicate.isProvidedByLibrary(reference.getClassName())) {
+        // these are the references to library classes which need to be checked at runtime
+        continue;
+      }
+      if (needToKeepFieldsAndMethods.contains(reference)) {
+        // these need to be kept in order to check abstract methods are implemented and declared
+        // super class fields are present
+        // TODO (trask) if this is provided by javaagent then can remove methods since the whole
+        //  class will be resolved at runtime
+        //  or another option is to resolve from helper classes first, see
+        //  HelperReferenceWrapper.Factory.create(String)
+        continue;
+      }
+      i.remove();
+    }
+  }
+
+  private Set<Reference> getHelperClassesWithLibrarySuperType() {
+    Set<Reference> helperClassesWithLibrarySuperType = new HashSet<>();
+    for (Map.Entry<String, Reference> entry : references.entrySet()) {
+      Reference reference = entry.getValue();
+      if (instrumentationClassPredicate.isInstrumentationClass(reference.getClassName())
+          && hasLibrarySuperType(reference.getClassName())) {
+        helperClassesWithLibrarySuperType.add(reference);
+      }
+    }
+    return helperClassesWithLibrarySuperType;
+  }
+
+  private void addSuperClasses(@Nullable String className, Set<Reference> superClasses) {
+    if (className != null && !className.startsWith("java.")) {
+      Reference reference = references.get(className);
```

Comment:
Nit: wouldn't be a bit safer to check if `reference != null` instead?

Reply:
changed the condition, I think more clear now that it can't (or shouldn't) be null